File: store.lmdb/src/main/java/mb/pie/store/lmdb/LMDBStore.java
package mb.pie.store.lmdb;

import mb.pie.api.*;
import org.lmdbjava.Dbi;
import org.lmdbjava.DbiFlags;
import org.lmdbjava.Env;
import org.lmdbjava.Txn;

import java.io.File;
import java.nio.ByteBuffer;

public class LMDBStore implements Store {
    public static final long defaultMaxDbSize = 1024 * 1024 * 1024; // 1 GiB
    public static final int defaultMaxReaders = 64; // 64 reader threads

    /**
     * Sets the store of this builder to the [LMDBStore], stored at given [envDir], with parameters [maxDbSize] determining the maximum
     * database size and [maxReaders] determining the maximum concurrent readers.
     */
    public static PieBuilder withLMDBStore(PieBuilder pieBuilder, File envDir, long maxDbSize, int maxReaders) {
        pieBuilder.withStoreFactory((logger) -> new LMDBStore(logger, envDir, maxDbSize, maxReaders));
        return pieBuilder;
    }

    /**
     * Sets the store of this builder to the [LMDBStore], stored at given [envDir], with default maximum database size and number of readers.
     */
    public static PieBuilder withLMDBStore(PieBuilder pieBuilder, File envDir) {
        return withLMDBStore(pieBuilder, envDir, defaultMaxDbSize, defaultMaxReaders);
    }


    private final Env<ByteBuffer> env;
    private final Dbi<ByteBuffer> input;
    private final Dbi<ByteBuffer> output;
    private final Dbi<ByteBuffer> taskObservability;
    private final Dbi<ByteBuffer> taskRequires;
    private final Dbi<ByteBuffer> callersOf;
    private final Dbi<ByteBuffer> callersOfValues;
    private final Dbi<ByteBuffer> resourceRequires;
    private final Dbi<ByteBuffer> requireesOf;
    private final Dbi<ByteBuffer> requireesOfValues;
    private final Dbi<ByteBuffer> resourceProvides;
    private final Dbi<ByteBuffer> providerOf;
    private final Logger logger;

    public LMDBStore(Logger logger, File envDir) {
        this(logger, envDir, defaultMaxDbSize, defaultMaxReaders);
    }

    public LMDBStore(Logger logger, File envDir, long maxDbSize, int maxReaders) {
        envDir.mkdirs();
        this.env = Env.create()
            .setMapSize(maxDbSize)
            .setMaxReaders(maxReaders)
            .setMaxDbs(11)
            .open(envDir);
        this.input = env.openDbi("input", DbiFlags.MDB_CREATE);
        this.output = env.openDbi("output", DbiFlags.MDB_CREATE);
        this.taskObservability = env.openDbi("taskObservability", DbiFlags.MDB_CREATE);
        this.taskRequires = env.openDbi("taskRequires", DbiFlags.MDB_CREATE);
        this.callersOf = env.openDbi("callersOf", DbiFlags.MDB_CREATE, DbiFlags.MDB_DUPSORT);
        this.callersOfValues = env.openDbi("callersOfValues", DbiFlags.MDB_CREATE);
        this.resourceRequires = env.openDbi("resourceRequires", DbiFlags.MDB_CREATE);
        this.requireesOf = env.openDbi("requireesOf", DbiFlags.MDB_CREATE, DbiFlags.MDB_DUPSORT);
        this.requireesOfValues = env.openDbi("requireesOfValues", DbiFlags.MDB_CREATE);
        this.resourceProvides = env.openDbi("resourceProvides", DbiFlags.MDB_CREATE);
        this.providerOf = env.openDbi("providerOf", DbiFlags.MDB_CREATE);
        this.logger = logger;
    }

    @Override public void close() {
        env.close();
    }

    @Override public StoreReadTxn readTxn() {
        final Txn<ByteBuffer> txn = env.txnRead();
        return new LMDBStoreTxn(env, txn, false, logger,
            input,
            output,
            taskObservability,
            taskRequires,
            callersOf,
            callersOfValues,
            resourceRequires,
            requireesOf,
            requireesOfValues,
            resourceProvides,
            providerOf
        );
    }

    @Override public StoreWriteTxn writeTxn() {
        final Txn<ByteBuffer> txn = env.txnWrite();
        return new LMDBStoreTxn(env, txn, true, logger,
            input,
            output,
            taskObservability,
            taskRequires,
            callersOf,
            callersOfValues,
            resourceRequires,
            requireesOf,
            requireesOfValues,
            resourceProvides,
            providerOf
        );
    }

    @Override public void sync() {
        env.sync(false);
    }

    @Override public String toString() {
        return "LMDBStore()";
    }
}

File: store.lmdb/src/main/kotlin/mb/pie/store/lmdb/LMDBStore.kt
package mb.pie.store.lmdb

import mb.pie.api.*
import org.lmdbjava.*
import java.io.File

/**
 * Sets the store of this builder to the [LMDBStore], stored at given [envDir], with parameters [maxDbSize] determining the maximum
 * database size and [maxReaders] determining the maximum concurrent readers.
 */
@JvmOverloads
fun PieBuilder.withLMDBStore(envDir: File, maxDbSize: Int = 1024 * 1024 * 1024, maxReaders: Int = 1024): PieBuilder {
  this.withStore { logger -> LMDBStore(logger, envDir, maxDbSize, maxReaders) }
  return this
}

typealias EnvB = Env<Buf>
typealias DbiB = Dbi<Buf>
typealias TxnB = Txn<Buf>

class LMDBStore(
  private val logger: Logger,
  envDir: File,
  maxDbSize: Int = 1024 * 1024 * 1024, // 1 GiB
  maxReaders: Int = 64 // 64 concurrent read threads
) : Store {
  private val env: EnvB
  private val input: DbiB
  private val output: DbiB
  private val taskRequires: DbiB
  private val callersOf: DbiB
  private val callersOfValues: DbiB
  private val resourceRequires: DbiB
  private val requireesOf: DbiB
  private val requireesOfValues: DbiB
  private val resourceProvides: DbiB
  private val providerOf: DbiB


  init {
    envDir.mkdirs()
    env = Env.create()
      .setMapSize(maxDbSize.toLong())
      .setMaxReaders(maxReaders)
      .setMaxDbs(10)
      .open(envDir)
    input = env.openDbi("input", DbiFlags.MDB_CREATE)
    output = env.openDbi("output", DbiFlags.MDB_CREATE)
    taskRequires = env.openDbi("taskRequires", DbiFlags.MDB_CREATE)
    callersOf = env.openDbi("callersOf", DbiFlags.MDB_CREATE, DbiFlags.MDB_DUPSORT)
    callersOfValues = env.openDbi("callersOfValues", DbiFlags.MDB_CREATE)
    resourceRequires = env.openDbi("resourceRequires", DbiFlags.MDB_CREATE)
    requireesOf = env.openDbi("requireesOf", DbiFlags.MDB_CREATE, DbiFlags.MDB_DUPSORT)
    requireesOfValues = env.openDbi("requireesOfValues", DbiFlags.MDB_CREATE)
    resourceProvides = env.openDbi("resourceProvides", DbiFlags.MDB_CREATE)
    providerOf = env.openDbi("providerOf", DbiFlags.MDB_CREATE)
  }

  override fun close() {
    env.close()
  }


  override fun readTxn(): StoreReadTxn {
    val txn = env.txnRead()
    return LMDBStoreTxn(env, txn, false, logger,
      inputDb = input,
      outputDb = output,
      taskRequiresDb = taskRequires,
      callersOfDb = callersOf,
      callersOfValuesDb = callersOfValues,
      resourceRequiresDb = resourceRequires,
      requireesOfDb = requireesOf,
      requireesOfValuesDb = requireesOfValues,
      resourceProvidesDb = resourceProvides,
      providerOfDb = providerOf
    )
  }

  override fun writeTxn(): StoreWriteTxn {
    val txn = env.txnWrite()
    return LMDBStoreTxn(env, txn, true, logger,
      inputDb = input,
      outputDb = output,
      taskRequiresDb = taskRequires,
      callersOfDb = callersOf,
      callersOfValuesDb = callersOfValues,
      resourceRequiresDb = resourceRequires,
      requireesOfDb = requireesOf,
      requireesOfValuesDb = requireesOfValues,
      resourceProvidesDb = resourceProvides,
      providerOfDb = providerOf
    )
  }

  override fun sync() {
    env.sync(false)
  }


  override fun toString(): String {
    return "LMDBStore"
  }
}
